sky/spot/recovery_strategy.py

```diff
@@ -142,39 +142,57 @@ def _launch(self, max_retry=3, raise_on_failure=True) -> Optional[float]:
                 # code.
                 logger.info(
                     f'Failed to launch the spot cluster with error: {e}')
-
-            cluster_status, _ = backend_utils.refresh_cluster_status_handle(
-                self.cluster_name, force_refresh=True)
-            if cluster_status == global_user_state.ClusterStatus.UP:
+                if max_retry is not None and retry_cnt >= max_retry:
+                    # Retry forever if max_retry is None.
+                    if raise_on_failure:
+                        with ux_utils.print_exception_no_traceback():
+                            raise exceptions.ResourcesUnavailableError(
+                                'Failed to launch the spot cluster after '
+                                f'{max_retry} retries.') from e
+                    else:
+                        return None
+                gap_seconds = backoff.current_backoff()
+                logger.info(
+                    f'Retrying to launch the spot cluster in {gap_seconds:.1f} '
+                    'seconds.')
+                time.sleep(gap_seconds)
+                continue
+
+            status = None
+            retry_launch = False
+            job_checking_retry_cnt = 0
+            while (status is None or status in [
+                    job_lib.JobStatus.INIT,
+                    job_lib.JobStatus.PENDING,
+            ]) and job_checking_retry_cnt < spot_utils.MAX_JOB_CHECKING_RETRY:
+                job_checking_retry_cnt += 1
+                cluster_status, _ = backend_utils.refresh_cluster_status_handle(
```

Comment:
This function can raise an error which is uncaught:

https://github.com/skypilot-org/skypilot/issues/1142#issuecomment-1235705061

Can we / is it correct to catch it so that we can go into `retry_launch`?

Reply:
I feel like this PR is more for the case where the cluster is preempted just after it is launched. The referred issue and comment is related to another issue that involves with the certification of gcloud being overwritten by the ray autoscaler. I am planning to have another PR to fix that. Wdyt?